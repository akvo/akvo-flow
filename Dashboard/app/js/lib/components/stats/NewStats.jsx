import React from 'react';
import PropTypes from 'prop-types';

export default class Stats extends React.Component {
  state = {
    startDate: '',
    endDate: '',
  };

  onChange = e => {
    this.setState({ [e.target.id]: e.target.value });
  };

  onSubmit = () => {
    const { startDate, endDate } = this.state;

    const data = {
      startDate: startDate.length ? new Date(startDate).toISOString() : null,
      endDate: endDate.length
        ? new Date(endDate).toISOString()
        : new Date().toISOString(),
    };

    this.props.generateReport(data);
  };

  render() {
    const { strings } = this.props;
    return (
      <div id="stats-page">
        <h2>{strings.generateStats}</h2>

        <p>{strings.formTimeFrame}</p>

        <div className="date-picker">
          <label htmlFor="startDate">
            <p>{strings.startDate}:</p>

            <input
              type="date"
              id="startDate"
              value={this.state.startDate}
              onChange={this.onChange}
              className="datePicker"
              style={{ width: '95%' }}
            />
          </label>

          <label htmlFor="endDate">
            <p>{strings.toDate}:</p>

            <input
              type="date"
              id="endDate"
              value={this.state.endDate}
              onChange={this.onChange}
              className="datePicker"
              style={{ width: '95%' }}
              min={this.state.startDate}
            />
          </label>
        </div>

        <button onClick={this.onSubmit} className="standardBtn" type="button">
          {strings.downloadStats}
        </button>
      </div>
    );
  }
}

Stats.propTypes = {
  generateReport: PropTypes.func.isRequired,
  strings: PropTypes.object.isRequired,
};
