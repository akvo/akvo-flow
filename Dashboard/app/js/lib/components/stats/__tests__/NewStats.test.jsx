import React from 'react';
import { render, cleanup, fireEvent } from '@testing-library/react';
import NewStats from '../NewStats';
import '@testing-library/jest-dom/extend-expect';

// https://github.com/testing-library/react-testing-library#suppressing-unnecessary-warnings-on-react-dom-168
// eslint-disable-next-line no-console
const originalError = console.error;
beforeAll(() => {
  // eslint-disable-next-line no-console
  console.error = (...args) => {
    if (/supports * the "act"/.test(args[0])) {
      return;
    }
    originalError.call(console, ...args);
  };
});

afterAll(() => {
  // eslint-disable-next-line no-console
  console.error = originalError;
});

describe('New Stats Tests', () => {
  afterEach(cleanup);

  const props = {
    strings: {
      generateStats: 'Generate form submission stats',
      formTimeFrame: 'Form submission time frame',
      startDate: 'Start date',
      toDate: 'To date',
      downloadStats: 'Download Stats',
    },
    generateReport: jest.fn(),
  };

  it('+++ renders snapshots', () => {
    const wrapper = render(<NewStats {...props} />);

    expect(wrapper.container).toMatchSnapshot();
  });

  it('+++ handles inputs and submit forms', () => {
    const wrapper = render(<NewStats {...props} />);

    // get start and end date input
    const startDate = wrapper.getByLabelText('Start date:');
    const endDate = wrapper.getByLabelText('To date:');

    // select date on inputs
    fireEvent.change(startDate, { target: { value: '2019-10-01' } });
    fireEvent.change(endDate, { target: { value: '2019-10-10' } });

    // click on submit button
    fireEvent.click(wrapper.getByText('Download Stats'));

    // expectations
    expect(props.generateReport).toHaveBeenCalledTimes(1);
    expect(props.generateReport).toHaveBeenCalledWith({
      // be called with ISO strings
      startDate: '2019-10-01T00:00:00.000Z',
      endDate: '2019-10-10T00:00:00.000Z',
    });
  });
});
