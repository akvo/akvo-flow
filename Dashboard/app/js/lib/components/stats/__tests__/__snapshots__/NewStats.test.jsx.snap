// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`New Stats Tests +++ renders snapshots 1`] = `
<div>
  <div
    id="stats-page"
  >
    <h2>
      Generate form submission stats
    </h2>
    <p>
      Form submission time frame
    </p>
    <div
      class="date-picker"
    >
      <label
        for="startDate"
      >
        <p>
          Start date
          :
        </p>
        <input
          class="datePicker"
          id="startDate"
          style="width: 95%;"
          type="date"
          value=""
        />
      </label>
      <label
        for="endDate"
      >
        <p>
          To date
          :
        </p>
        <input
          class="datePicker"
          id="endDate"
          min=""
          style="width: 95%;"
          type="date"
          value=""
        />
      </label>
    </div>
    <button
      class="standardBtn"
      type="button"
    >
      Download Stats
    </button>
  </div>
</div>
`;
